src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -384,18 +423,18 @@ private Delegate[] MakeGetters(IRow input, Func<int, bool> activeOutput)
                 var outputCache = new OutputCache();
                 var activeOutputColNames = _parent.Outputs.Where((x, i) => activeOutput(i)).ToArray();
 
-                var valueGetters = new List<Delegate>();
+                var valueGetters = new Delegate[_parent.Outputs.Length];
                 for (int i = 0; i < _parent.Outputs.Length; i++)
                 {
                     if (activeOutput(i))
                     {
                         var type = TFTensor.TypeFromTensorType(_parent.TFOutputTypes[i]);
                         _host.Assert(type == _parent.OutputTypes[i].ItemType.RawType);
                         var srcTensorGetters = GetTensorValueGetters(input);
-                        valueGetters.Add(Utils.MarshalInvoke(MakeGetter<int>, type, input, i, srcTensorGetters, activeOutputColNames, outputCache));
+                        valueGetters[i] = Utils.MarshalInvoke(MakeGetter<int>, type, input, i, srcTensorGetters, activeOutputColNames, outputCache);
```

Comment:
>valueGetters[i] [](start = 24, length = 15)

am curious about this bug in the getter -- could you kindly elaborate a bit on this ? .. is this some artifact of the activateOutput() call above..

Reply:
This bug surfaced when I added the new unit test. It happens in the following situation:

- Create a TensorFlowTransform that computes two outputs, say A and B.
- Create another transform/learner that only uses B as input.
- When we try to cursor over the data, the activeOutput predicate will say that output 0 is not active and output 1 is active, thus returning an array of length 1.
- When we try to get the getter of column B, we do so using its index, which is 1 (the index of column A is 0 and the index of column B is 1). So we try to access the getters array which is of length 1, at index 1 which is out of bounds...

The fix was to always create an array with length equal to the number of output columns in the transform, but populate just the indices where the active columns are.



---
In reply to: [216469314](https://github.com/dotnet/machinelearning/pull/857#discussion_r216469314) [](ancestors = 216469314)